src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -5318,7 +5305,7 @@ int RGWRados::Object::complete_atomic_modification(const DoutPrefixProvider *dpp
     return 0;
   }
 
-  string tag = (state->tail_tag.length() > 0 ? state->tail_tag.to_str() : state->obj_tag.to_str());
+  string tag = state->get_obj_refcount_tag();
```

Comment:
this had been using `state->tail_tag` which comes from `RGW_ATTR_TAIL_TAG`. i see that `copy_obj()` calls `attrs.erase(RGW_ATTR_TAIL_TAG)` in the ref-counting path. can we set that to `tag + '\0'` instead?

Reply:
wait, that should be happening already in copy_obj() -> write_meta() -> prepare_atomic_modification()

Reply:
The PR kept the state->tail_tag unmodified and it is still used as the ref_tag on the SRC object.
The copy_obj() erase the TAIL_TAG on the target destination as an indication that it is was produced by copy and in that case will use the ID_TAG as ref_tag (which is wrong, as we used something else )

Reply:
> The copy_obj() erase the TAIL_TAG on the target destination as an indication that it is was produced by copy and in that case will use the ID_TAG as ref_tag (which is wrong, as we used something else )

copy_obj() erases RGW_ATTR_TAIL_TAG, but later adds it back in prepare_atomic_modification(). it looks like that should correctly store the new ref_tag in both RGW_ATTR_ID_TAG and RGW_ATTR_TAIL_TAG. are you sure that's not happening?

Reply:
the target object _should_ be atomic here, and i'd expect `s->object->set_atomic()` in [RGWCopyObj::verify_permission()](https://github.com/ceph/ceph/blob/cac7dcd6/src/rgw/rgw_op.cc#L5397) to make it so

Reply:
ok, the problem is that we're calling s->src_object->copy_object() which calls copy_obj() with src_object's RGWObjectCtx. s->object->set_atomic() is setting the atomic flag in the RGWObjectCtx of s->object

cc @dang, i think this was a regression from https://github.com/ceph/ceph/pull/44616. prior to that, we use the same `s->obj_ctx` for both objects